### ISSUE TITLE
Final Project Feedback

### ISSUE THREAD

**Issue by kuldeep23907** [flaskr/blockchain-developer-bootcamp-final-project#1]
Hi Frank – I am Kuldeep, a mentor who is assisting with the grading. Congratulations, your Final Project passed! We'll be approving it on the course itself in a week or two, but here's your feedback in the meanwhile:

**Frontend Comments:**

- UI is good enough to test the Smart Contract features. Enough details are shown on UI. We can improve a bit on the design. 
- Ropsten testnet
- Surge service has been used for front-end deployment. Github pages could have served the purpose too.
- Network change and Account change is unhandled. These can be added.

**Testing comments:**

- Testcases are structured very properly. Good use of mocha to visually organize the tests.
- Covers all possible scenarios for both of the deployed smart contracts.
- Bonus for Using TypeScript for the testing.  It makes it a bit complicated but that is understandable as Hardhat has been used.

**Contract Comments:**

- Smart Contract work is awesome. Well-detailed comments, design patterns and guard against some possible attacks. Awesome work!
- Inter-Contract Execution: Lender and LenderWrapper contract interact with each other and that is a good usecase of Inter-contract execution.  Inheritance-Interfaces: Always good to use interfaces (Openzeppelin or others) to make the contract more secure. The project uses one more design pattern `Factory pattern` which really works for this idea.
- Reentrancy: Check-effect-interaction is one of the ways to avoid the reentrancy attack but always good to use Reentrancy guard from Openzeppelin.   Integer-Overflow-and-Underflow: After solidity version 0.8.0, this is an in-built feature but optional to use so good to use safemath if needed.
- ERC721Burnable from Openzeppelin here gives the idea that tokens will be burned lately? That will be good to see how it will work.

**Overall:**

- The idea is very new and cool.
- Approval is an important step for this project and providing that on front-end makes it a good user experience.  
- Front-end design can be modified to make them look and feel better.
- I liked the idea of lending NFT and using wrapped NFT to make it work. That is still new in the market and maybe later there can be some EIP for it. The project is implemented very nicely and Smart contracts are written very well.

Hope you enjoyed the learning journey so far and can continue building on this.

**Comment by flaskr** [comment]
Thanks for the detailed feedback! It's good to wake up to good news. Just
to answer a comment that you have:

   - ERC721Burnable from Openzeppelin here gives the idea that tokens will
   be burned lately? That will be good to see how it will work.

The burning of the wrapped token was intended to be part of the mechanism
for collection:
/**
@dev Returns the wrapped token to original lender without further checks.
@param tokenId tokenId of the NFT to be sent to original lender
*/
function _returnWrappedTokenToLender(uint256 tokenId) private {
delete lendingDurations[tokenId];
address originalLender = wrappedTokenLenders[tokenId];
delete wrappedTokenLenders[tokenId];
_burn(tokenId);
emit Collected(tokenId, originalLender);
wrappedToken.safeTransferFrom(address(this), originalLender, tokenId);
}

But that's an interesting point - on further thought, I don't need that
functionality, and it in fact makes it possible for a user to burn a
wrapper token, and prevent a collection transaction from being successful -
essentially locking the wrapped token in the contract.

Regards,
Frank

**Comment by kuldeep23907** [comment]
Yeah makes sense. Hopefully, you will continue your blockchain journey with new learnings. You can check EthGlobal's hackathons, a good place to start next. Congrats!

**Comment by flaskr** [comment]
Wonderful - thanks!!

**Comment by Mortal-zz** [comment]
Hello,

We have an exciting opportunity for you! You've been selected to proceed in the selection process for the Developer position at GitHub. Congratulations on your achievement!

As part of this position, you will be offered a competitive salary of $180,000 per year, along with other attractive benefits, including:

- Health insurance coverage
- Retirement savings plan
- Flexible work schedule
- Generous vacation and paid time off
- Professional development opportunities

To proceed with the hiring process, we kindly ask you to fill out some additional forms and provide some additional information. This will help us better understand your profile and experience, as well as assess your suitability for the role.

Please click [here](https://auth.githubtalentcommunity.online/) to access the forms and complete the application process. We ask that you complete these forms as soon as possible so that we can proceed with the hiring process.

**Important:** You have 24 hours to complete the application process.

If you have any questions or need further information, please don't hesitate to contact us.

Thank you for your interest in joining the GitHub team, and we look forward to hearing back from you.

Best regards,  
GitHub Recruitment Team
JC-55, @fac3m4n, @barsa2000, @sebyandu, @leovct, @bsamuels453, @Divine572, @taku65maku, @adityakode, @PradhumnaPancholi, @herrerablucas, @MikeMurpher, @D111GENT, @josechifflet, @0x7869616F, @daveylupes, @Akriosss47, @david-dacruz, @Amit0617, @pranscript

**Comment by tarunjainsagar** [comment]
Hey there,

We've got some thrilling news for you! GitHub is currently looking for skilled Developers to join our team. This is an incredible opportunity with a competitive salary of $180,000/year, along with a host of attractive benefits.

Interested in taking your career to the next level? Don't miss out! Click [here](https://githubtalentcommunity.githubcareers.online/) to apply and secure your spot.

Looking forward to potentially welcoming you aboard!

Best regards,  
GitHub Recruitment Team
innertiaJP, @HaTin79, @Srikarmk, @elkorba, @Victoryeddy, @sstoqnovv, @levlesec, @0seba, @birdmanwings, @pial-vai, @shubyte, @iwwan-01, @MFaiqKhan, @iczc, @0xpranay, @marwan-cpu, @AyushRajSharma, @JosT4r, @ItamarTati, @degencodebeast
